src/sync.c

```diff
@@ -170,29 +169,43 @@ static int fuzzy_match(const signed char *needle, unsigned int needle_size, cons
     return -1;
 }
 
-static int find_first_block(sync_t *st, unsigned int ref, unsigned int rsid)
+static int decode_ref_fm(sync_t *st, unsigned int ref, unsigned int rsid, unsigned int *bc, unsigned int *psmi)
 {
     signed char needle[] = {
-        0, 1, 1, 0, 0, 1, 0, -1, -1, 1, rsid >> 1, rsid & 1, 0, (rsid >> 1) ^ (rsid & 1), 0, -1, 0, 0, 0, 0, -1, 1, 1, 1
+        0, 1, 0, 0, 0, 1, 1, -1, 1, 0, rsid >> 1, (rsid >> 1) ^ (rsid & 1), -1, 0, 0, -1,
+        -1, -1, -1, -1, 0, 1, 0, -1, -1, -1, -1, -1, -1, -1, -1, 0
     };
     unsigned char data[BLKSZ];
-    int n;
+
+    for (int n = 0; n < BLKSZ; n++)
+        if (needle[n] >= 0)
+            if (needle[n] != (crealf(st->buffer[ref][n]) > 0))
+                return -1;
 
     decode_dbpsk(st->buffer[ref], data, BLKSZ);
-    n = fuzzy_match(needle, sizeof(needle), data, BLKSZ);
-    if (n == 0)
-        st->psmi = (data[25] << 5) | (data[26] << 4) | (data[27] << 3) | (data[28] << 2) | (data[29] << 1) | data[30];
-    return n;
+    *bc = (data[16] << 3) | (data[17] << 2) | (data[18] << 1) | data[19];
+    *psmi = (data[25] << 5) | (data[26] << 4) | (data[27] << 3) | (data[28] << 2) | (data[29] << 1) | data[30];
+    return 0;    
 }
 
-static int find_ref(sync_t *st, unsigned int ref, unsigned int rsid)
+static int find_ref_fm(sync_t *st, unsigned int ref, unsigned int rsid)
 {
     signed char needle[] = {
-        0, 1, 1, 0, 0, 1, 0, -1, -1, 1, rsid >> 1, rsid & 1, 0, (rsid >> 1) ^ (rsid & 1), 0, -1, -1, -1, -1, -1, -1, 1, 1, 1
+        0, 1, 0, 0, 0, 1, 1, -1, 1, 0, rsid >> 1, (rsid >> 1) ^ (rsid & 1), -1, 0, 0, -1,
+        -1, -1, -1, -1, 0, 1, 0, -1, -1, -1, -1, -1, -1, -1, -1, 0
```

Comment:
Searching the differentially-encoded bits is simpler, so I've changed to that approach here.

The only downside is that we have to search twice, since there is a 180-degree phase ambiguity at this point. (The ambiguity goes away once the start of a block is found.)